Modules/_multiprocessing/posixshmem.c

```diff
@@ -0,0 +1,724 @@
+problem (issue 9586) and I copied their solution here.
+ref: http://bugs.python.org/issue9586
+
+Note that in /Developer/SDKs/MacOSX10.4u.sdk/usr/include/sys/semaphore.h,
+SEM_FAILED is #defined as -1 and that's apparently the definition used by
+Python when building. In /usr/include/sys/semaphore.h, it's defined
+as ((sem_t *)-1).
+*/
+#ifdef __APPLE__
+    #undef  SEM_FAILED
+    #define SEM_FAILED ((sem_t *)-1)
+#endif
+
+/* POSIX says that a mode_t "shall be an integer type". To avoid the need
+for a specific get_mode function for each type, I'll just stuff the mode into
+a long and mention it in the Xxx_members list for each type.
+ref: http://www.opengroup.org/onlinepubs/000095399/basedefs/sys/types.h.html
+*/
+
+typedef struct {
+    PyObject_HEAD
+    char *name;
+    long mode;
+    int fd;
+} SharedMemory;
+
+
+// FreeBSD (and perhaps other BSDs) limit names to 14 characters. In the
+// code below, strings of this length are allocated on the stack, so
+// increase this gently or change that code to use malloc().
+#define MAX_SAFE_NAME_LENGTH  14
+
+
+/* Struct to contain an IPC object name which can be None */
+typedef struct {
+    int is_none;
+    char *name;
+} NoneableName;
+
+
+/*
+      Exceptions for this module
+*/
+
+static PyObject *pBaseException;
+static PyObject *pPermissionsException;
+static PyObject *pExistentialException;
+
+
+#ifdef POSIX_IPC_DEBUG
+#define DPRINTF(fmt, args...) fprintf(stderr, "+++ " fmt, ## args)
+#else
+#define DPRINTF(fmt, args...)
+#endif
+
+static char *
+bytes_to_c_string(PyObject* o, int lock) {
+/* Convert a bytes object to a char *. Optionally lock the buffer if it is a
+   bytes array.
+   This code swiped directly from Python 3.1's posixmodule.c by Philip S.
+   The name there is bytes2str().
+*/
+    if (PyBytes_Check(o))
+        return PyBytes_AsString(o);
+    else if (PyByteArray_Check(o)) {
+        if (lock && PyObject_GetBuffer(o, NULL, 0) < 0)
+            /* On a bytearray, this should not fail. */
+            PyErr_BadInternalCall();
+        return PyByteArray_AsString(o);
+    } else {
+        /* The FS converter should have verified that this
+           is either bytes or bytearray. */
+        Py_FatalError("bad object passed to bytes2str");
+        /* not reached. */
+        return "";
+    }
+}
+
+static void
+release_bytes(PyObject* o)
+    /* Release the lock, decref the object.
+   This code swiped directly from Python 3.1's posixmodule.c by Philip S.
+   */
+{
+    if (PyByteArray_Check(o))
+        o->ob_type->tp_as_buffer->bf_releasebuffer(NULL, 0);
+    Py_DECREF(o);
+}
+
+
+static int
+random_in_range(int min, int max) {
+    // returns a random int N such that min <= N <= max
+    int diff = (max - min) + 1;
+
+    // ref: http://www.c-faq.com/lib/randrange.html
+    return ((int)((double)rand() / ((double)RAND_MAX + 1) * diff)) + min;
+}
+
+
+static
+int create_random_name(char *name) {
```

Comment:
This function is not necessary, see below.

Reply:
Agree with this and Neil Schemenauer's comment that this would be better implemented in Python than in C. I didn't do so in `posix_ipc` because I there don't have the luxury of a Python wrapper.

Reply:
Moved to Python thanks to Neil's patch in GH-11816.